Add strategy document

iHiD opened the pull request:
I've added a document listing Exercism's current working strategy. 

I would appreciate feedback on:
- Spelling/grammar mistakes
- Clarifications if something isn't clear
- Anything that I've discussed with you but missed off here.

Please note that I do not want to use this PR as a place to debate the merits of the strategy itself - an issue on exercism/exercism is the appropriate place for that :)

**Maintainers: When you're ready, please approve this, but leave me to merge it as I need to add it on the website at the same time.**

Thank y'all! 💙

verdammelt commented on pages/strategy.md:
```diff
@@ -0,0 +1,58 @@
+
+### Track Anatomy Project (TAP)
+
+Some solutions are fun to mentor, some are extremely challenging. Foe example, solutions that require maths lead to mentors trying to tutor someone on algebra, not on coding. Other solutions can be solved in such a variety of ways that the mentor ends up debating the right algorithm, not teaching the idioms of the language.
```
s/Foe/For

iHiD replied:
```suggestion
Some solutions are fun to mentor, some are extremely challenging. For example, solutions that require maths lead to mentors trying to tutor someone on algebra, not on coding. Other solutions can be solved in such a variety of ways that the mentor ends up debating the right algorithm, not teaching the idioms of the language.
```

iHiD replied:
Thanks.

kytrinyx commented on pages/strategy.md:
```diff
@@ -0,0 +1,58 @@
+
+Some solutions are fun to mentor, some are extremely challenging. For example, solutions that require maths lead to mentors trying to tutor someone on algebra, not on coding. Other solutions can be solved in such a variety of ways that the mentor ends up debating the right algorithm, not teaching the idioms of the language.
+
+TAP will fix this by providing a framework to Track Maintainers, which they can work through in order to determine the optimal series of core exercises, and sensible related side exercises. In our prototype Track (Ruby), the result of TAP saw a average reduction in waiting time from one week to half a day, and reduced the queue from hundreds to ~30 at any time. It also saw a doubling in the amoun ...
```
> doubling in the amount of submissions

_doubling in the **number** of submissions_

iHiD replied:
```suggestion
TAP will fix this by providing a framework to Track Maintainers, which they can work through in order to determine the optimal series of core exercises, and sensible related side exercises. In our prototype Track (Ruby), the result of TAP saw a average reduction in waiting time from one week to half a day, and reduced the queue from hundreds to ~30 at any time. It also saw a doubling in the number of submissions per week and dramatically increased mentor retention, with the most active weekly number mentors of any track. Encoding this into a framework and applying it to all tracks should have a dramatic affect on mentor enjoyment.
```

kytrinyx commented on pages/strategy.md:
```diff
@@ -0,0 +1,58 @@
+
+### Remove the barrier to mentoring
+
+We currently have a deliberate barrier to mentoring with a wall that you have to sign up through. We have put this in place because we want to try and ensure a level of quality control and therefore want to gate-keep and ensure that people who are mentoring feel like they are up to the taks. However, as we add automation, documentation, and a more integrated community, we will be able to remove t ...
```
> We currently have a deliberate barrier to mentoring with a wall that you have to sign up through.

How about:

_We currently have a deliberate barrier to mentoring, requiring mentors to sign up._

> we want to try and ensure

_try **to** ensure_ I think?

> feel like they are up to the taks

_feel like they are up to the **task**

> rather than something for the elite.

Maybe: _rather than something exclusive._

SleeplessByte commented on pages/strategy.md:
```diff
@@ -0,0 +1,58 @@
+Exercism is a large complex project with thousands of contributors working across dozens of projects, all trying to achieve ambitious goals. It would quickly become overwhelming if we were all working towards making Exercism perfect, so to keep things manageable, we periodically determine the critical next step for Exercism to achieve its goals and then plan that into a strategy. Our current goal ...
```
As this is a long read, I would personally prefer that the goals are separate from this big paragraph:

```suggestion
Exercism is a large complex project with thousands of contributors working across dozens of projects, all trying to achieve ambitious goals. It would quickly become overwhelming if we were all working towards making Exercism perfect, so to keep things manageable, we periodically determine the critical next step for Exercism to achieve its goals and then plan that into a strategy.

Our _current_ goal is to make Exercism scalable.

We expect our next three goals to be:
- to make Exercism financially sustainable,
- to make Exercism more accessible to beginners,
- and to intentionally grow Exercism for the first time.
```
------------

Exercism is a large complex project with thousands of contributors working across dozens of projects, all trying to achieve ambitious goals. It would quickly become overwhelming if we were all working towards making Exercism perfect, so to keep things manageable, we periodically determine the critical next step for Exercism to achieve its goals and then plan that into a strategy.

Our _current_ goal is to make Exercism scalable.

We expect our next three goals to be:
- to make Exercism financially sustainable,
- to make Exercism more accessible to beginners,
- and to intentionally grow Exercism for the first time.

SleeplessByte commented on pages/strategy.md:
```diff
@@ -0,0 +1,58 @@
+Exercism is a large complex project with thousands of contributors working across dozens of projects, all trying to achieve ambitious goals. It would quickly become overwhelming if we were all working towards making Exercism perfect, so to keep things manageable, we periodically determine the critical next step for Exercism to achieve its goals and then plan that into a strategy. Our current goal ...
+
+We present this as an Objective, which will aim to achieve with a Strategy, supported by a series of Tactics. This is commonly known as the OST framework.
```
```suggestion
We present this as an **Objective**, which will aim to achieve with a **Strategy**, supported by a series of **Tactics.** This is commonly known as the OST framework.
```

I suggest you highlight these words.

SleeplessByte commented on pages/strategy.md:
```diff
@@ -0,0 +1,58 @@
+
+**Make Exercism scalable**
+
+At the time of writing, Exercism has 8,000 new submissions per week. Our mentors currently cannot keep up with the amount of submissions submitted. When they do keep up, the number of submissions almost immediately doubles, as students stick with a Language Track. We are growing by 300-400 new users per day. Many find Exercism too hard to use, but in the future, when that is fixed, we will expect ...
```
Same as the long read all the way at the start of the document, I would separate the concrete goal from the paragraph, like so (including Katrina's change):

```suggestion
At the time of writing, Exercism has 8,000 new submissions per week. Our mentors currently cannot keep up with the number of submissions submitted. When they do keep up, the number of submissions almost immediately doubles, as students stick with a Language Track. We are growing by 300-400 new users per day. Many find Exercism too hard to use, but in the future, when that is fixed, we will expect to see the number of submissions double again. We have never actively marketed Exercism. All 300,000 members have come through word of mouth. When we start to market Exercism, we expect to see an increase in usage by two orders of magnitude. 

Our most important goal at this stage is being able to give our current students a timely mentoring experience, and be able to cope with further growth without overwhelming mentors.
```

I also don't now if it's preferable to combine so many different subjects into one block. You're talking about scalability, but also that many find exercism too hard which targets the progression lines and learning curve. 

I think it might be good to remove that all together and instead write that "those track that have been undergoing restructure via the project track anatomy" have seen significant increase in the number of submissions, or something similar in your words. That said, since in the **Tactics** part you elaborate on this, I think this paragraph works fine without the entire sentence.

iHiD replied:
For clarity, I mean "hard to use" as in things like downloading the CLI rather than finding the exercises too hard.

SleeplessByte replied:
Now that you say it, it makes sense. This is not clear from the wording 🗡 😅

SleeplessByte commented on pages/strategy.md:
```diff
@@ -0,0 +1,58 @@
+
+**Acquire and retain more mentors by making mentoring rewarding and enjoyable**
+
+Currently, mentoring can be both fulfilling and frustrating. It can be very rewarding to help a student that wants to learn. On the flipside, it can be repetitive, students can be rude, there are few positive feedback loops, and it can feel too much like a job rather than a hobby. In order to retain our existing mentors, and to acquire new ones, we must make mentoring more enjoyable and more rewa ...
```
I think a semicolon separated list reads better here:
```markdown
On the flipside: it can be repetitive, students can be rude; there are few positive feedback loops; and it 
can feel too much like a job rather than a hobby
```

SleeplessByte replied:
>  We must make our mentors feel like that are part of something bigger them them,

 We must make our mentors feel like that are part of something bigger **than** them

SleeplessByte replied:
> A good parallel is wikipedia, where

A good parallel is *Wikipedia*, where

SleeplessByte replied:
>  straight-forward but (strangely) rewarding

 straight-forward **; and** (strangely) rewarding

iHiD replied:
I'm not a huge fan of semi-colons here - I think commas do the job ok.

SleeplessByte commented on pages/strategy.md:
```diff
@@ -0,0 +1,58 @@
+
+Some solutions are fun to mentor, some are extremely challenging. For example, solutions that require maths lead to mentors trying to tutor someone on algebra, not on coding. Other solutions can be solved in such a variety of ways that the mentor ends up debating the right algorithm, not teaching the idioms of the language.
+
+The Track Anatomy Project will fix this by providing a framework to Track Maintainers, which they can work through in order to determine the optimal series of core exercises, and sensible related side exercises. In our prototype Track (Ruby), the result of the Track Anatomy Project saw a average reduction in waiting time from one week to half a day, and reduced the queue from hundreds to ~30 at a ...
```
> by providing a framework to Track Maintainers

Does one "work through a framework"? Perhaps calling it a tool might better?

> The Track Anatomy Project **fixes** this by providing a **tool** to Track Maintainers, which they can **use** to determine...

Then again, calling it a framework at the end of the paragraph makes complete sense. 

> to all tracks should have a dramatic affect on mentor enjoyment.

Might be good to be explicit that this is positive:

> to all tracks should have a **dramatically positive** affect on mentor enjoyment.

SleeplessByte commented on pages/strategy.md:
```diff
@@ -0,0 +1,58 @@
+
+The majority of submissions are common. It is rare for a student to submit a solution that no-one else has ever submitted. The Track Anatomy Project aims to ensure this is the case, guiding tracks to have core exercises that are usually solved with one of a few common approaches. Automated Mentoring Support aims to utilize this consistency to provide automated approval and/or feedback to students ...
+- **Reduce student frustration:** We give students a quick feedback loop. If a solution is optimal, let them continue. If it fails in a known way, tell them and let them improve it. This keeps students getting frustrated with waiting for their perfectly-ok solutions to be approved and therefore with the mentors who they perceive to be the source of the slowness.
+- **Dramatically reduce the size of the queue:** Our prototyping has shown that for simple exercises (e.g. TwoFer) we will be able to provide automated mentoring in 90%+ of cases. For more complex exercises, he percentage will be smaller but we expect to be able to provide automated mentoring on 50% of exercises, and on all optimal, or close-to-optimal, solutions.
```
```suggestion
- **Dramatically reduce the size of the queue:** Our prototyping has shown that for simple exercises (e.g. TwoFer) we will be able to provide automated mentoring in 90%+ of cases. For more complex exercises, the percentage will be smaller but we expect to be able to provide automated mentoring on 50% of exercises, and on all optimal, or close-to-optimal, solutions.
```

iHiD commented:
@SleeplessByte @kytrinyx Thank you both so much 💙 

Katrina and I discussed this last night and agreed that there was a mix of abstraction levels in the tactics that made reading it unclear. These were my concluding thoughts:

> I wrote this as if we're standing at the start of this objective's journey, but we're reading it a few months into the journey. So we've made progress with some of the points (track anatomy, automated mentoring), but we've not worked on others (e.g. documentation). When we read it now, we see those ones we've worked on as more defined and solid, but if we went back to when I'm trying to write this from, they would all have been equally fuzzy (we had no idea what Maud would actually do, for example, just that we needed her to do it)
>
> I would rather they were all fuzzy, because I intend this to be read as if it was written at the start, not half-way through.

I've updated the document to try and make it all a bit more "fuzzy" and added "updates" to the three points we have progress with so as not to lose detail that might be valuable for the reader. I'd appreciate another scan over to see if this is better :)

rpottsoh commented on pages/strategy.md:
```diff
@@ -0,0 +1,73 @@
+
+## Objective
+
+**Make Exercism scalable**
```
```suggestion
###Make Exercism scalable
```
Later in the DOC you you switch to this pattern from `**<say somthing>**`.

iHiD replied:
@rpottsoh Thanks :) Yeah - this was deliberate. The reason is that there's only one objective and only one strategy, so visually having the two headings look a bit "off" to me. Once it's merged on the website, I'll check again to see if I think that makes sense or whether I should shift to what you suggest.

rpottsoh commented on pages/strategy.md:
```diff
@@ -0,0 +1,73 @@
+
+## Strategy
+
+**Acquire and retain more mentors by making mentoring rewarding and enjoyable**
```
```suggestion
###Acquire and retain more mentors by making mentoring rewarding and enjoyable
```
If I have missed the point for the change in formatting mid-stream then disregard.

kytrinyx commented on pages/strategy.md:
```diff
@@ -0,0 +1,73 @@
+
+### Automatically analyze common solutions
+
+The majority of submissions are common. It is rare for a student to submit a solution that no-one else has ever submitted. As part of restructing the track, we aim to enforce this as a contraint, guiding tracks to have core exercises that are usually solved with one of a few common approaches. We aim to utilize this consistency to provide automated approval and/or feedback to students for the maj ...
```
> we aim to enforce this as a constraint

It's not clear what "this" points to.

kytrinyx commented on pages/strategy.md:
```diff
@@ -0,0 +1,73 @@
+
+### Automatically analyze common solutions
+
+The majority of submissions are common. It is rare for a student to submit a solution that no-one else has ever submitted. As part of restructing the track, we aim to enforce this as a constraint, guiding tracks to have core exercises that are usually solved with one of a few common approaches. We aim to utilize this consistency to provide automated approval and/or feedback to students for the ma ...
```
I think my previous comment got lost in history somewhere (I'm not seeing it, anyway).

> We aim to enforce this as constraint

It's not clear what "this" points to.

rpottsoh commented on pages/strategy.md:
```diff
@@ -0,0 +1,73 @@
+### Automatically analyze common solutions
+
+The majority of submissions are common. It is rare for a student to submit a solution that no-one else has ever submitted. As part of restructing the track, we aim to enforce this as a constraint, guiding tracks to have core exercises that are usually solved with one of a few common approaches. We aim to utilize this consistency to provide automated approval and/or feedback to students for the ma ...
+- **Reduce student frustration:** We give students a quick feedback loop. If a solution is optimal, let them continue. If it fails in a known way, tell them and let them improve it. This prevents students from getting frustrated with waiting for their perfectly-ok solutions to be approved and therefore with the mentors who they perceive to be the source of the slowness.
```
```suggestion
- **Reduce student frustration:** We give students a quick feedback loop. If a solution is optimal, let them continue. If it fails in a known way, tell them and let them improve it. This prevents students from becoming frustrated waiting for their perfectly-OK solutions from being approved reducing the perception of mentors being the source of slow approvals.
```
I think this reads a little easier.

iHiD replied:
I've tweaked it slightly, but fundamentally what you wrote. Thanks :)

iHiD commented:
Thanks everyone. I'm merging this now so I can reference it in tomorrow's mentor update. More than happy for people to PR any extra changes :)
